disperser/apiserver/validation.go

```diff
@@ -0,0 +1,239 @@
+
+	data := req.GetData()
+	blobSize := len(data)
+	// The blob size in bytes must be in range [1, maxBlobSize].
+	if blobSize > maxBlobSize {
+		return nil, fmt.Errorf("blob size cannot exceed 2 MiB")
+	}
+	if blobSize == 0 {
+		return nil, fmt.Errorf("blob size must be greater than 0")
+	}
+
+	if len(req.GetCustomQuorumNumbers()) > 256 {
+		return nil, errors.New("invalid request: number of custom_quorum_numbers must not exceed 256")
+	}
+
+	quorumConfig, err := s.updateQuorumConfig(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get quorum config: %w", err)
+	}
+
+	if len(req.GetCustomQuorumNumbers()) > int(quorumConfig.QuorumCount) {
+		return nil, errors.New("invalid request: number of custom_quorum_numbers must not exceed number of quorums")
+	}
+
+	seenQuorums := make(map[uint8]struct{})
+	// The quorum ID must be in range [0, 254]. It'll actually be converted
+	// to uint8, so it cannot be greater than 254.
+	for i := range req.GetCustomQuorumNumbers() {
+
+		if req.GetCustomQuorumNumbers()[i] > 254 {
+			return nil, fmt.Errorf("invalid request: quorum_numbers must be in range [0, 254], but found %d", req.GetCustomQuorumNumbers()[i])
+		}
+
+		quorumID := uint8(req.GetCustomQuorumNumbers()[i])
+		if quorumID >= quorumConfig.QuorumCount {
+			return nil, fmt.Errorf("invalid request: the quorum_numbers must be in range [0, %d], but found %d", s.quorumConfig.QuorumCount-1, quorumID)
+		}
+
+		if _, ok := seenQuorums[quorumID]; ok {
+			return nil, fmt.Errorf("invalid request: quorum_numbers must not contain duplicates")
+		}
+		seenQuorums[quorumID] = struct{}{}
+
+	}
+
+	// Add the required quorums to the list of quorums to check
+	for _, quorumID := range quorumConfig.RequiredQuorums {
+		if _, ok := seenQuorums[quorumID]; ok {
+			return nil, fmt.Errorf("invalid request: quorum_numbers should not include the required quorums, but required quorum %d was found", quorumID)
+		}
+		seenQuorums[quorumID] = struct{}{}
+	}
+
+	if len(seenQuorums) == 0 {
+		return nil, fmt.Errorf("invalid request: the blob must be sent to at least one quorum")
+	}
+
+	params := make([]*core.SecurityParam, len(seenQuorums))
+	i := 0
+	for quorumID := range seenQuorums {
+		params[i] = &core.SecurityParam{
+			QuorumID:              core.QuorumID(quorumID),
+			AdversaryThreshold:    quorumConfig.SecurityParams[quorumID].AdversaryThreshold,
+			ConfirmationThreshold: quorumConfig.SecurityParams[quorumID].ConfirmationThreshold,
+		}
+		err = params[i].Validate()
+		if err != nil {
+			return nil, fmt.Errorf("invalid request: %w", err)
+		}
+		i++
+	}
+
+	header := core.BlobRequestHeader{
+		BlobAuthHeader: core.BlobAuthHeader{
+			AccountID: req.AccountId,
+		},
+		SecurityParams: params,
+	}
+
+	blob := &core.Blob{
+		RequestHeader: header,
+		Data:          data,
+	}
+
+	return blob, nil
+}
+
+func (s *DispersalServer) getAccountRate(origin, authenticatedAddress string, quorumID core.QuorumID) (*PerUserRateInfo, string, error) {
```

Comment:
The rate limiting is generally not in the category of request validation, but in access control category (authentication, authorization, rate limiting etc)

Reply:
Do you want me to move this to an `authentication` file?

Reply:
Or access control? To cover both auth and rate limiting